core-rust/state-manager/src/store/jmt_gc.rs

```diff
@@ -153,29 +163,38 @@ impl StateTreeGc {
                     // only after its children, in case this process is interrupted half-way
                     // and need to be resumed).
                     StaleTreePart::Subtree(subtree_root_key) => {
-                        Box::new(iterate_dfs_post_order(database.deref(), subtree_root_key))
+                        Box::new(iterate_dfs_post_order(database, subtree_root_key))
                     }
                 };
-                for key in part_keys {
-                    deleted_nodes.push(key);
-                    // Periodically rotate the collected buffer of node keys to delete:
-                    if deleted_nodes.len() == DELETED_NODE_BUFFER_MAX_LEN {
-                        info!("Flushing a full delete buffer at version {}", state_version);
-                        database.batch_delete_node(deleted_nodes.iter());
-                        deleted_nodes.clear();
-                    }
-                }
+                deleted_nodes.extend(part_keys);
             }
-            deleted_state_versions.push(state_version);
         }
 
-        // Delete the last collected batch of keys, and then delete the processed "stale tree parts" records:
+        // Delete the last collected batch of keys:
         info!("Flushing the last buffer ({} deletes)", deleted_nodes.len());
-        database.batch_delete_node(deleted_nodes.iter());
-        database.batch_delete_stale_tree_part(deleted_state_versions.iter());
+        delete_nodes_and_update_state_history(database, &deleted_nodes, to_state_version);
     }
 }
 
+/// A helper method for flushing the GC's partial DB update batches.
+///
+/// Note: the order of the operations is important for non-trivial reasons - see comments.
```

Comment:
This is fine, although a part of me would rather we did this in one `rw_context`, I'm not sure we have a good reason not to, and that way we don't have to worry about ordering so much? :)

Reply:
yeah, but from previous comment:

> It's a little weird to me that we combine the deletion of the stale tree notifications with the updating of the status.

(which is also valid remark)

Doing this under one `rw_context` would combine 3 semi-related items into one `StateManagerDatabase`-level call. This is because we never exposed any notion of "transaction" (or "== DB write batch") outside `StateManagerDatabase`. And it looks like "the right" solution here would require exactly that non-trivial piece of infra.

Of the 2 imperfect alternatives, I prefer to stay with the current state in this PR - just for the sake of conventions that I see around:
- the methods in DB traits are quite atomic / single-responsibility;
- other long-running DB processes (e.g. backfills) also use the right ordering of DB operations (to be resilient to Node restarts)

(ofc my opinion above assumes that the current impl does avoid that previous pitfall - if we find any problem with it, it will be clear signal that "yeah we better put that under one `do_these_3_things()` DB call")